wgsl/index.bs

```diff
@@ -4070,10 +4069,121 @@ Issue: in Vulkan, builtin variables occoupy I/O location slots counting toward l
 
 #### User Data Attribute TODO #### {#user-data-attributes}
 
-TODO: User data attributes must not be of [=bool=] type or contain a [=bool=] type.
+User-defined data can be passed as input to the start of a pipeline, passed
+between stages of a pipeline or output from the end of a pipeline.
+User-defined IO must not be passed to [=compute=] shader entry points.
+User-defined IO must be an [=IO-shareable=] type, but must not be or contain a
+[=bool=] or [=array=] type.
+All user defined IO must be assigned a location (See [[#input-output-locations]]).
 
-#### Input-output Locations TODO #### {#input-output-locations}
-TODO: *Stub*. Location-sizing of types, non-overlap among variables referenced within an entry point static call tree.
+#### Input-output Locations #### {#input-output-locations}
+
+Each location represents 16-bytes of storage.
+The size of elements are determined using the *SizeOf* column in [[#alignment-and-size]].
+For example, a four-element vector of floating-point values consumes a single location.
+
+Locations are specified via the `[[location(x)]]` attribute.
+The attribute must only be applied to entry point function parameters, entry
+point return values or structure members.
+The attribute must only be applied to [=numeric scalar=] or [=numeric vector=] types.
+The attribute must not be used with the [=compute=] shader stage.
+The locations assigned to structure members must be monotonically increasing.
+Locations are not packed between different members of a structure.
+Each structure member must be assigned a unique location.
+
+Every user-defined input and output must have a fully specified set of locations.
+That is, there must be no parameter, return value or structure member missing a
+location unless it is a builtin variable (See [[#builtin-inputs-outputs]]).
+
+For a given entry point, the locations of the return value are distinct from
+the locations of the function parameters.
+Within each set of locations, there must be no overlap.
+
+Note: the number of available locations for an entry point is defined by the WebGPU API.
+
+<div class='example applying location attribute' heading='Applying location attributes'>
+  <xmp>
+    struct A {
+      [[location(0)]] x : f32;
+      // Despite locations being 16-bytes, x and y cannot share a location
+      [[location(1)]] y : f32;
+    };
+
+    struct B {
+      a : A; // consumes locations 0 and 1
```

Comment:
if this PR is addressing the nested structures, it should state the requirements differently, i.e. "each member is either a struct (with the rule applied to members), or a built-in, or has a defined location"

Reply:
I've rewritten the rules to try and clarify how it should work with nested structures.